database/7-things-a-developer-should-know-(about-database)/post.md

```diff
@@ -0,0 +1,155 @@
+
+I dedicate this post to Michel Vayssade. 15 years ago, your courses mixed high abstraction with down-to-earth concretion, in your demanding and unadorned style. What a joy to reap again the benefit of them in understanding databases !
+
+## TL;DR
+
+Full-stack and back-end developers, you probably don't know this - and it can get out of trouble:
+
+- `pg_stat_activity` view shows executing queries, `pg_stat_statements` shows executed queries, queries can be logged;
+- use a pool; make sure when scaling you do not reach max_connections;
+- activate `default_transaction_read_only` and `statement_timeout` in your sql client in production;
+- never login to production database OS/container: use a sql client instead;
+- if `pg_stat_activity` shows queries waiting for lock, use the lock view to find which ones got them;
+- when a SQL query has started, it will run until completion - doesn't matter if the client is gone;
+- use `pg_terminate_backend` to stop a query, and be ready for AUTOVACUUM.
+
+## Why should I care ?
+
+99% of the time, developer doesn't have to care about database. As long as they practice the basics they learned in initial training (a sound - and versioned - data model, consistency with transaction, and some index for performance), they don't have to think much about it. The database is a nice black box. As for me, a web developer, I enjoy learning database internals for fun: it gives me the kind of thrill I only got in systems programming courses.
+
+Over the years, I've come to realize that some bits of what I learned from internals were useful for anyone. Actually, these bits may come and bite you harshly in production. So, instead of laughing under my breath when these things happen to others, I'm going to share them now.
+
+My first idea was to produce "an emergency kit", but thanks to Tom Kyte (from the introduction of his book "Effective Oracle by Design") I've included preventive knowledge, especially concurrency. As we have to know so many things, I've kept the list short. For the same reasons, I only cover PostgreSQL, being the most used relational database these days by our clients.
+
+## Preventive healthcare
+
+### Know your database
+
+> First things first: before going live, make your database observable.
+
+When bad things will happen in production (and they will), it will too late to realize you don't know what is actually happening. From the project's onset, in the [walking skeleton](https://wiki.c2.com/?WalkingSkeleton) - the first time code is deployed on a remote environment, you should know which queries are under execution in the database.
+
+A native PG view does exactly that : [pg_stat_activity](https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-ACTIVITY-VIEW). It mentions which user is executing which query on which database, and the status of this query: is it waiting for the disk, is it waiting for a lock ? 
+
+Make your job easier :
```

Comment:
[ the colon is never preceded by a white space](url)https://www.sussex.ac.uk/informatics/punctuation/colonandsemi/colon#:~:text=But%20first%20please%20note%20the,have%20been%20taught%20in%20school.

Reply:
Corrigé, dommage qu'il n'y ai pas de règle dans https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md

Reply:
Corrigé, dommage qu'il n'y ai pas de règle dans https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md